#64 refresh password views

akostromin14 opened the pull request:
- Create TokenManager
- Refresh PasswordViews
- Refresh serializer for PasswordView
- Sort imports in used files

akostromin14 committed:
remove shared tasks from utils

akostromin14 committed:
create basic TokenManager

akostromin14 committed:
remove old app

akostromin14 committed:
add invalid token error

akostromin14 committed:
refresh password recovery and password reset

akostromin14 committed:
update errors in PasswordRecoverySerializer

akostromin14 committed:
remove unnecessary code

akostromin14 committed:
update status code for not found

akostromin14 committed:
update errors to custom where it's possible

akostromin14 committed:
update manager except Error

akostromin14 committed:
Merge branch 'develop' of https://github.com/Project-Stage-Academy/UA-1131_Forum-PS into #64-refresh-password-views

akostromin14 committed:
remove old imports

YanZhylavy reviewed (commented):
Great job, thank you for caring about code quality! But I noticed a small issue - if an email that doesn't exist in the database or an incorrect email is entered during login, the server crashes because the error raised by CustomUser.get_user(email=email) is not handled. It seems like this issue existed before and you didn't touch LoginView, but since you're currently refactoring authentication views, it's worth taking this into account.

mehalyna commented on authentication/views.py:
```diff
@@ -104,44 +114,89 @@ def post(self, request):
+
+        payload = TokenManager.get_payload(jwt_token)
+        user_id = payload.get('user_id')
+        if user_id is None:
+            return Response({'error': Error.INVALID_TOKEN.msg}, status=Error.INVALID_TOKEN.status)
```
It's good that you're checking if `user_id` is `None`. But it looks like your `TokenManager.get_payload()` method doesn't include robust error handling for cases where the token is invalid, expired, or tampered with. Consider also handling other potential issues with the token, such as signature verification failures

akostromin14 replied:
The fact of the matter if that SimpleJWT returns 1 exception for all errors (I include screenshots). The TokenError inherits from Python Exception.
![image](https://github.com/Project-Stage-Academy/UA-1131_Forum-PS/assets/76693557/91b89f9a-73d5-4d59-a3f4-7c3feef7612b)
![image](https://github.com/Project-Stage-Academy/UA-1131_Forum-PS/assets/76693557/50233d74-6b01-41d9-bc00-f5e27897c212)

mehalyna commented on authentication/views.py:
```diff
@@ -104,44 +114,89 @@ def post(self, request):
+        user_id = payload.get('user_id')
+        if user_id is None:
+            return Response({'error': Error.INVALID_TOKEN.msg}, status=Error.INVALID_TOKEN.status)
+        user = CustomUser.get_user(user_id=user_id)
```
Consider adding error handling for cases where no user matches the provided `user_id`, which could happen if the user record has been deleted or the `user_id`in the token payload is invalid

akostromin14 replied:
Fixed

mehalyna commented on forum/managers.py:
```diff
@@ -0,0 +1,39 @@
+    def get_payload(cls, token) -> dict:
+        """Returns payload for token"""
+
+        decoded_token = cls.__get_decoded_token(token)
```
This method assumes that every token passed to it will be a valid `AccessToken`. Consider cases where you might need to decode different types of tokens (e.g., `RefreshToken`). You might either handle different token types in this method or create separate methods for different token types if their handling diverges significantly

akostromin14 replied:
I choose the second method to create separate methods for different token types, to avoid nested try-except blocks.

akostromin14 committed:
handle user not exist &  separate token types

handle user not exits error and create separate methods for different token types
add typing for manager

akostromin14 committed:
Merge branch 'develop' of https://github.com/Project-Stage-Academy/UA-1131_Forum-PS into #64-refresh-password-views
